File: frontend/ui/src/app/types/portal.ts
export interface PortalOIDCProvider {
  name: string;
  loginPath: string;
  spInitiated: boolean;
}

export type RegistrationMode = 'enabled' | 'hidden' | 'disabled';

export interface PortalLoginConfig {
  registration: RegistrationMode;
  turnstileSiteKey?: string;
  oidcGithubEnabled: boolean;
  oidcGoogleEnabled: boolean;
  oidcMicrosoftEnabled: boolean;
  oidcGenericEnabled: boolean;
  oidcProviders: PortalOIDCProvider[];
}

export interface Portal {
  // Whether the request host matches a custom app domain. Used to drop Distr-specific branding (logo, website
  // links) on custom domains, even when no branding assets are configured.
  customDomain: boolean;
  pageTitle?: string;
  faviconUrl?: string;
  logoUrl?: string;
  // Absent when the instance has no SUPPORT_EMAIL configured, in which case support must not be offered as a link.
  supportEmail?: string;
  loginConfig: PortalLoginConfig;
}

File: api/portal.go
package api

import "github.com/distr-sh/distr/internal/env"

type PortalResponse struct {
	CustomDomain bool              `json:"customDomain"`
	PageTitle    *string           `json:"pageTitle,omitempty"`
	FaviconUrl   *string           `json:"faviconUrl,omitempty"`
	LogoUrl      *string           `json:"logoUrl,omitempty"`
	LoginConfig  PortalLoginConfig `json:"loginConfig"`
}

type PortalLoginConfig struct {
	Registration         env.RegistrationMode `json:"registration"`
	TurnstileSiteKey     *string              `json:"turnstileSiteKey,omitempty"`
	OIDCGithubEnabled    bool                 `json:"oidcGithubEnabled"`
	OIDCGoogleEnabled    bool                 `json:"oidcGoogleEnabled"`
	OIDCMicrosoftEnabled bool                 `json:"oidcMicrosoftEnabled"`
	OIDCGenericEnabled   bool                 `json:"oidcGenericEnabled"`
	OIDCProviders        []PortalOIDCProvider `json:"oidcProviders,omitempty"`
}

type PortalOIDCProvider struct {
	Name        string `json:"name"`
	LoginPath   string `json:"loginPath"`
	SPInitiated bool   `json:"spInitiated"`
}
